File: Template/python/BeverageMakerTemplate.py
from abc import ABC, abstractmethod
from typing import final

class BeverageMaker(ABC):
    @abstractmethod
    def brew(self) -> None:
        pass

    @abstractmethod
    def addCondiments(self) -> None:
        pass

    def boilWater(self) -> None:
        print("Boiling water")

    def pourInCup(self) -> None:
        print("Pouring into cup")

    @final
    def makeBeverage(self) -> None:
        self.boilWater()
        self.brew()
        self.pourInCup()
        self.addCondiments()
        
class TeaMaker(BeverageMaker):
    def brew(self) -> None:
        print("Steeping the tea")

    def addCondiments(self) -> None:
        print("Adding lemon")
        
class CoffeeMaker(BeverageMaker):
    def brew(self) -> None:
        print("Dripping coffee through filter")

    def addCondiments(self) -> None:
        print("Adding sugar and milk")

File: Template/ts/BeverageMakerTemplate.ts
export abstract class BeverageMaker {
    // This method should be final, but TypeScript does not support it
    public makeBeverage(): void {
        this.boilWater();
        this.brew();
        this.pourInCup();
        this.addCondiments();
    }

    public boilWater(): void {
        console.log('Boiling water');
    }

    public pourInCup(): void {
        console.log('Pouring in cup');
    }

    public abstract brew(): void;

    public abstract addCondiments(): void;
}

export class TeaMaker extends BeverageMaker {
    public brew(): void {
        console.log('Steeping the tea');
    }

    public addCondiments(): void {
        console.log('Adding lemon');
    }
}

export class CoffeeMaker extends BeverageMaker {
    public brew(): void {
        console.log('Dripping coffee through filter');
    }

    public addCondiments(): void {
        console.log('Adding sugar and milk');
    }
}